### DISCUSSION TITLE
Creating a Manifold compliant object from surfaces.

### DISCUSSION THREAD

**Discussion by pyenthu** [elalish/manifold#877] [Q&A]
Basic question. Just started using the manifold library and finding that a number of objects I am creating are not compatible with this library because of the manifold test. Just wondering if there is any library or guidance on creating a solid body from surfaces that is manifold compliant?
I tried the convexhull library from threejs but it has a limitation that if there is a concave undulation on the surface it will draw a straight line. 
I am just trying to avoid reconstructing complex NURBS surfaces for manifold compliance using triangles.
As you can see the blue line is the surface profile, the straight mesh fillers are based on manifold-3d and NURBS and convex hull.

<img width="405" alt="Screenshot 2024-07-25 at 2 58 53 AM" src="https://github.com/user-attachments/assets/0103366c-e0bc-4918-ac82-7f511e38f3f4">

**Comment by elalish** [comment]
It's pretty hard to guess what you're trying to do from that image - it looks more 2D than 3D. As for convex hull, first we have our own in this library, and second, filling in concavity is the very definition of what a convex hull does. 

How would you plan to "avoid reconstructing complex NURBS surfaces for manifold compliance using triangles"? This library uses triangles, not NURBS, so that reconstruction is pretty much a given, right? Part of why I built this library is because the NURBS-based CAD kernels have a lot of trouble making anything manifold, and that's due to NURBS themselves. 

Turning non-manifold mesh data into a manifold is non-trivial. That's why the guarantees of this library are a big deal - you'll never have to go through that process with anything this library produces. On the input side though, you will have to do the work. What is your actual input data? Is it close enough that `MeshGL.Merge()` helps?

**Reply by pyenthu** [reply]
I am trying, and will post some code soon. I am trying to see how I can "close" a couple of surfaces (NURBS) to form a Manifold.
It is a non-trival problem but very useful to do. I also, did not know about MeshGL. Let me look. 
Will post soon.

**Reply by pyenthu** [reply]
Please find my first attempt to create a compliant solid. It plots fine but is not Manifold. 
All I have done so far is to ensure that the triangles are counterclockwise. Does not work yet.
{
    "name": "ManifoldError",
    "message": "Not manifold",
    "stack": "ManifoldError: Not manifold\n    at new ManifoldError (http://localhost:5173/node_modules/manifold-3d/manifold.js?v=4a1c0f75:8:9190)\n    at new Module.Manifold (http://localhost:5173/node_modules/manifold-3d/manifold.js?v=4a1c0f75:8:11640)\n    at createMFMesh (http://localhost:5173/src/routes/geo/geocube/mftrial.ts?t=1721898051904:30:12)\n    at create_fragment (http://localhost:5173/src/routes/geo/geocube/SceneEx.svelte?t=1721898051904:1166:15)\n    at init (http://localhost:5173/node_modules/.vite/deps/chunk-H5NYQPBQ.js?v=4a1c0f75:2199:35)\n    at new SceneEx (http://localhost:5173/src/routes/geo/geocube/SceneEx.svelte?t=1721898051904:1661:3)\n    at createProxiedComponent (http://localhost:5173/node_modules/svelte-hmr/runtime/svelte-hooks.js?v=4a1c0f75:341:9)\n    at new ProxyComponent (http://localhost:5173/node_modules/svelte-hmr/runtime/proxy.js?v=4a1c0f75:242:7)\n    at new Proxy<SceneEx> (http://localhost:5173/node_modules/svelte-hmr/runtime/proxy.js?v=4a1c0f75:349:11)\n    at Array.create_default_slot_5 (http://localhost:5173/src/routes/geo/geocube/ThrelteEx.svelte?t=1721898051904:68:12)",
    "code": 2
}
```
// Create a mesh from scratch
const front = [
  [0, 0, 0],
  [1000, 0, 0],
  [1000, 1000, 0],
  [0, 1000, 0],
];

const back = [
  [1000, 0, 1000],
  [0, 0, 1000],
  [0, 1000, 1000],
  [1000, 1000, 1000],
];

const sqFront = [0, 1, 2, 3],
  sqBack = [4, 5, 6, 7],
  sqRight = [1, 4, 7, 2],
  sqLeft = [0, 3, 6, 5],
  sqTop = [3, 2, 7, 6],
  sqBottom = [0, 1, 4, 5];

const sqToTri = (sq: number[]) => [sq[0], sq[1], sq[2], sq[2], sq[3], sq[0]];

const vertices = [sqFront, sqRight, sqBack, sqLeft, sqTop, sqBottom].map(
  sqToTri
);
console.log("CONVERTED TO TRIANGLES ", vertices);

//Create a manifold cube
export function createMFMesh() {
  const mesh = new Mesh({
    numProp: 3,
    vertProperties: new Float32Array([...front.flat(), ...back.flat()]),
    triVerts: new Uint32Array([...vertices.flat()]),
  });

  mesh.merge();
  let mf = new Manifold(mesh);
  //   console.log("MANIFOLD !!", mf);
  let threegeom: BufferGeometry = mesh2geometry(mesh);
  threegeom.computeVertexNormals();
  //   return new BoxGeometry
  return threegeom;
}

```

**Reply by pyenthu** [reply]
One question, is there a way to use your warp function to create a parametric/implicit surface? Is there a good library that would do a NURBS kind of morphing? That might work with you library? 
Because it is going to be VERY difficult to create a Manifold mesh manually. I have been trynig the whole day. Even adding two driangles to a cube is tough.

**Reply by pca006132** [reply]
No, warp can only change the coordinates of points of an existing manifold object. What specific kind of morphing you want? We did not try to combine this with other libraries that does NURBS, so probably no recommendation here. Depending on use cases and precision requirements, you may be able to get away with sdf? Or you will need some custom solution.

Also, what do you want by adding two triangles to a cube? A cube is already a manifold, you cannot add faces to it without poking a hole first.

**Reply by elalish** [reply]
Yeah, use our `LevelSet()` function for implicit surfaces. It's easy to turn a NURBS swatch into triangles, the hard part is the joints between the NURBS. If you have six complete NURBS patches that join like the faces of a cube, then you could just do `Manifold::Cube().Refine(100).Warp(NURBfunc)`. For more general cases, even the $30,000/seat CAD packages have a tough time. 

Manifoldness just requires that every triangle's edge matches another triangles edge with the two verts in the opposite order. So simple, and yet not easy. If you can get it to where all the cracks in the model are smaller than some precision, then `MeshGL.Merge()` should take care of it.

**Reply by pyenthu** [reply]
Took your advice and created a simple level function. 
More familiar with THREE so worked with buffer geometry.
It works. For now, using a kriging function to do an "inverse Nurbs". Getting a ittle deviation from the surface when reconstructing this from the vertices of the buffer based on the trained model .
Reading about this a bit more they say that the RBF function is more accurate. Will try that next. 
It is reasonably fast. Looks like a highly viable option for what I am trying to do. Will be doing a lot of cutting and slicing.
I have seen people ask questions about this in other forums. I think it is a viable option to create a Manifold from Nurbs with a few lines of code.
Thanks for the guidance

```
import { surfToGeom } from "$geo/verbscene";
import { storeCubeDims } from "$geo/geocube/storesthrelte";
import { Manifold, geometry2mesh, mesh2geometry } from "$lib/manifold/threem";
import { BoxGeometry, BufferGeometry, SphereGeometry } from "three";
import { get } from "svelte/store";

async function getVerb() {
  const verb = await import("tuo-verb");
  return verb;
}
const verb = await getVerb();

// This is the kriging function
import { train, predict } from "@sakitam-gis/kriging";

//create a large swept NURBS surface using verb.geom.NurbsSurface
function createNurbsSurface() {
  var prof = new verb.geom.BezierCurve([
    [-get(storeCubeDims).length / 2, 800, 0],
    [0, 0, 0],
    [get(storeCubeDims).length / 2, 800, 0],
  ]);
  var rail = new verb.geom.BezierCurve([
    [0, 0, -get(storeCubeDims).width / 2],
    [0, 0, 0],
    [0, 0, get(storeCubeDims).width / 2],
  ]);

  const srf = new verb.geom.SweptSurface(prof, rail);
  const geom = surfToGeom(srf).translate(0, 0, -get(storeCubeDims).width / 2);
  geom.computeVertexNormals();
  return { geom, srf };
}

export function modelSurf(srf: any, zOffset = 0) {
  // loop through the surf with u and v values and console log their xyz values
  // These will be used for interpolation
  const x: number[] = [],
    y: number[] = [],
    z: number[] = [];

  // Z offset is required in some surfaces that are not centered at 0
  for (let u = 0; u <= 1; u += 0.05) {
    for (let v = 0; v <= 1; v += 0.05) {
      const pt = srf.point(u, v);
      x.push(pt[0]);
      y.push(pt[1]);
      z.push(pt[2] + zOffset);
    }
  }
  var model = "exponential",
    sigma2 = 2,
    alpha = 200;
  return train(y, x, z, model, sigma2, alpha);
}

//Create a manifold cube
export function surfToMFMesh(surf: any, zOffset = 0) {
  let fitModel = modelSurf(surf, zOffset);

  const box: BufferGeometry = new BoxGeometry(
    get(storeCubeDims).length,
    get(storeCubeDims).height,
    get(storeCubeDims).width,
    50,
    50,
    20
  ).translate(0, get(storeCubeDims).height / 2, 0);

  const refHeight = get(storeCubeDims).height;
  const heightAt = (x: number, y: number, z: number) => {
    let yNew = y + (predict(x, z, fitModel) * (refHeight - y)) / refHeight;
    return yNew;
  };

  // Modify box vertices with a radial function
  const pts = box.attributes.position.array;

  for (let i = 0; i < pts.length; i += 3) {
    pts[i + 1] = heightAt(pts[i], pts[i + 1], pts[i + 2]);
  }
  const mfSphere = new Manifold(
    geometry2mesh(new SphereGeometry(800, 50, 50).translate(500, 1000, 0))
  );
  let mfMan = new Manifold(geometry2mesh(box)).subtract(mfSphere);
  let geom = mesh2geometry(mfMan.getMesh());
  geom.computeVertexNormals();
  return { geom, mfMan };
}

export const surf = createNurbsSurface();
export const mfGeom = surfToMFMesh(
  surf.srf,
  -get(storeCubeDims).width / 2
).geom;

```

**Reply by elalish** [reply]
Awesome, glad to hear it! Would love to see some renders of what you're creating too. Feel free to add to #340 as well.
